File: spark/projects/streaming-example/src/main/scala/streamingexample/Window.scala
package com.thedataincubator.streamingexample

import org.apache.spark.streaming.StreamingContext
import scala.collection.mutable.Queue
import org.apache.spark.rdd.RDD
import org.apache.spark.streaming.Seconds

object Window {
  def job(ssc: StreamingContext) {
    val rddQueue = Queue[RDD[Int]]((1 to 100)
      .map( i => ssc.sparkContext.makeRDD(Seq(i))): _*
    )

    // Create the Socket Text Stream and use it do some processing
    val raw = ssc.queueStream(rddQueue, oneAtATime=true).map(_.toDouble)

    val value = raw.map(x => ("value" -> x))

    val lastTwoSum = raw.window(Seconds(2))
      .transform( rdd => ssc.sparkContext.makeRDD(Seq(rdd.sum)) )
      .map(x => ("lastTwoSum" -> x))

    val output = ssc.union(Seq(value, lastTwoSum)).print()

    ssc.start()
    ssc.awaitTermination()
  }
}

File: spark/projects/py-streaming-example/window.py
from pyspark import SparkConf, SparkContext
from pyspark.streaming import StreamingContext

def main():
    conf = SparkConf().setAppName("netcount")
    sc = SparkContext(conf=conf)
    ssc = StreamingContext(sc, 1)

    rdds = [sc.parallelize([i]) for i in xrange(100)]

    raw = ssc.queueStream(rdds, oneAtATime=True).map(float)
    value = raw.map(lambda x: ('value', x))
    last_two_sum = raw.window(2) \
                      .transform(lambda rdd: sc.parallelize([rdd.sum()])) \
                      .map(lambda x: ('sum', x))
    output = value.union(last_two_sum).pprint()

    ssc.start()
    ssc.awaitTermination()

if __name__ == '__main__':
    main()
